### PR TITLE
[security] fix 'assert' & 'xml' & 'md5' vulnerability in python files

### PR DESCRIPTION
**Pull request by AshburnLee**
Fix: https://github.com/intel/intel-xpu-backend-for-triton/issues/1961

## 1. Use raise exception instead of  `assert` in python files.
Test files are not included.

## 3. Fix MD5 issue
~~In Triton code, md5 is used for hash key generation, not for security purpose, so this PR adds `usedforsecurity=False` to solve severity issue found by Bandit~~
Replace md5 with sha256

### CODE REVIEW

**Comment by alexbaden** [comment]
Why is raising an exception better than assert here? 

Btw, it looks like upstream moved to sha256 vs md5: https://github.com/triton-lang/triton/blob/main/third_party/nvidia/backend/compiler.py#L115
though that was probably not a security concern

**Comment by pbchekin** [comment]
> Why is raising an exception better than assert here?

https://bandit.readthedocs.io/en/1.7.9/plugins/b101_assert_used.html:
"Assert is removed with compiling to optimised byte code (python -O producing *.opt-1.pyc files). This caused various protections to be removed. Consider raising a semantically meaningful error or AssertionError instead."

File: scripts/build_report.py [hunk 1]
```diff
@@ -48,7 +49,8 @@ def transform_df(df, param_cols, tflops_col, hbm_col, benchmark, compiler, tag):
         n: os.getenv(n.upper(), default="")
         for n in ["libigc1_version", "level_zero_version", "gpu_device", "agama_version"]
     }
-    assert host_info['gpu_device'], "Could not find GPU device description, was capture_device.sh called?"
+    if not host_info['gpu_device']:
```
Comment (pbchekin) [inline scripts/build_report.py:52]: Let's use consistent quotes within the file
```suggestion
    if not host_info["gpu_device"]:
```

File: third_party/intel/backend/compiler.py [hunk 2]
```diff
@@ -59,11 +59,11 @@ def __post_init__(self):
                                                  str(default_libdir / 'libsycl-spir64-unknown-unknown.bc'))
         object.__setattr__(self, 'extern_libs', tuple(extern_libs.items()))
         assert self.num_warps > 0 and (self.num_warps & (self.num_warps - 1)) == 0, \
-            "num_warps must be a power of 2"
+            "num_warps must be a power of 2"  # nosec B101
```
Comment (pbchekin) [inline third_party/intel/backend/compiler.py:62]: Why do we keep this `assert`? It is used for runtime validation and this is not a test.

**Comment by alexbaden** [comment]
I am loathe to diverge from upstream, particularly for performance sensitive code: https://github.com/triton-lang/triton/blob/main/third_party/nvidia/backend/compiler.py

**Comment by pbchekin** [comment]
> I am loathe to diverge from upstream, particularly for performance sensitive code: https://github.com/triton-lang/triton/blob/main/third_party/nvidia/backend/compiler.py

I understand, but nvidia backend is not our upstream, it's OK if our requirements for our code are stricter.

**Comment by alexbaden** [comment]
More strict in what sense? This is performance sensitive code because we want to reduce compilation latency as much as possible. To that end, the Nvidia backend is actually more strict as the asserts will be removed when max performance is desired.

**Comment by etiotto** [comment]
This is our code so we can change it without incurring into merge conflicts. I think that replacing asserts with exceptions is fine for error situations (rationale: https://github.com/intel/intel-xpu-backend-for-triton/pull/1962#issuecomment-2302126387). 

Do this changes actually cause performance problems ?

**Review by etiotto — APPROVED** [review]

**Comment by pbchekin** [comment]
> More strict in what sense?

Is this a rhetorical question? This PR literally fixes the issues found by Bandit. There are no Bandit scans in upstream.

> Do this changes actually cause performance problems ?

I don't think there is a performance impact. Running an external compiler (or any external process)  is much slower, for example. The question here is how important this runtime check and if we really want to skip in some cases.

**Comment by vlad-penkin** [comment]
> More strict in what sense? This is performance sensitive code because we want to reduce compilation latency as much as possible. To that end, the Nvidia backend is actually more strict as the asserts will be removed when max performance is desired.
I am not sure how to apply generic performance sensitive code to the changes in this PR. I second @pbchekin.

File: third_party/intel/backend/compiler.py (hunk 2 above)
Comment (AshburnLee) [reply third_party/intel/backend/compiler.py]: oh, I can raise `AssertionError` stead. 
Done, thanks~

File: scripts/build_report.py (hunk 1 above)
Comment (AshburnLee) [reply scripts/build_report.py]: Done

**Comment by AshburnLee** [comment]
Updated using `sha256` instead of `md5` with `usedforsecurity=False`

**Pull request merged**